### PR TITLE
test(codegen): #340 PR-A foundation — testharness extensions + FuzzSchema IR + rapid generators

### PR DESCRIPTION
**Pull request by invakid404**
Part 1 of 3 for #340. Foundation layer: testharness extensions + shared `FuzzSchema` IR + rapid generators + value walker. No oracle execution yet — PR-B adds dynamic, PR-C adds static.

## What lands

The `bamlfuzz` package and the testharness extensions PR-B and PR-C will both consume. Nothing here runs an oracle on its own.

### `adapters/common/codegen/internal/testharness/` (extended)

- `WriteTempBamlSrc(t, dir, files)` lays out a synthesized `baml_src/` under `t.TempDir()`.
- `WriteReplayArtifact(t, dir, name, content)` writes pretty-printed JSON for stable failure replays.
- `RunGoTestArgs` + `RunGoTestOptions` accept `Run` / `Fuzz` / `Tags` / `Race` / `Count` / `Env`. `GOWORK=off` is non-overridable. Legacy `RunGoTest` preserved as a wrapper so PR #341 callers compile unchanged.
- `CheckBamlSrcName` / `CheckReplayName` are pure path-validation helpers. Raw input is validated BEFORE `filepath.Clean` (rejecting raw separators, `.`/`..` segments, and Windows drive-relative prefixes) so canonicalization can't bypass the check.
- `hasWindowsDrivePrefix` is an OS-independent detector for `[A-Za-z]:` at the start of a name, so regression tests fire on every CI platform regardless of `filepath.VolumeName`'s OS behavior.

### `adapters/common/codegen/internal/bamlfuzz/` (new package)

- `schema.go` / `value.go` / `case.go` — pure data types: `FuzzSchema`, `FuzzValue`, `OracleCase`, `OracleMode` (`dynamic_three_way`, `static_prompt`), `CaseMetadata`. Includes `KindNull` as a first-class primitive alongside `KindString` / `KindInt` / `KindFloat` / `KindBool`.
- `selfref.go` — `AnalyzeGraph` separates `HasSelfRef` (class's own type tree reaches itself, including through `optional` / `list` / `map` wrappers) from `HasMutualCycle` (transitive cycle through other classes only). `RequiresDynamicSkip = HasSelfRef` only — mutual cycles are dynamic-friendly per the locked scope (the `TODO(upstream-self-ref)` placeholder marks the dynamic skip for removal when the upstream BAML TypeBuilder gains self-ref support).
- `normalize.go` — `Walk(schema, value)` produces stable JSON for mock LLM output and parser-normalized expected; records peak per-class recursion depth on each `s.depths[cls]++` (rather than post-walk, which had recorded zeros).
- `generator.go` — `DynamicSafeSchemaGen` (can emit mutual cycles via `MutualCycleProbability=0.15`, never self-ref) and `StaticSchemaGen` (probabilistic self-ref injection at `SelfRefProbability=0.15`). Injection REPLACES one of the drawn fields, preserving the `1..MaxFieldsPerClass` bound. `FindClass` / `FindEnum` lookups both fail fast on missing IDs.
- `pgregory.net/rapid v1.2.0` added as a direct test dependency.

## Bounds (scope D9)

- ≤4 classes, ≤3 enums
- 1–5 fields per class
- type depth ≤4
- list/map length 0–3
- value recursion cap 2 visits per class, terminated via `OptionalAbsent` / `[]` / `{}`

## Test coverage

- Unit tests pin IR JSON round-trip, the optional contract (present / absent / null → mock vs expected), `KindNull` walker behavior, map key lexicographic ordering, peak recursion-depth tracking, and the `AnalyzeGraph` self-ref vs mutual-cycle distinction.
- Rapid invariant tests prove: dynamic-safe never self-refs, stamped flags match `AnalyzeGraph`, schema/value bounds are respected, value walker terminates, mock→normalize round-trip is byte-equal to expected, dynamic-safe CAN produce mutual cycles (`TestDynamicSafeCanEmitMutualCycle`).
- testharness has direct tests for `WriteTempBamlSrc`, `WriteReplayArtifact`, and the path-validation helpers, including bypass classes (`sub/../case.json`, `shared/../main.baml`, `C:case.json`).
- `paths_test.go` covers `hasWindowsDrivePrefix` edge cases directly: `""`, `"C"`, `"C:"`, `"C:foo"`, `"c:foo"`, `"5:foo"`, `":foo"`, `"CC:foo"`, `"C:foo:bar"`, `"foo/C:bar"`.

## What this PR does NOT deliver (PR-B / PR-C will)

- Dynamic emitter (`FuzzSchema → bamlutils.DynamicOutputSchema`) → PR-B.
- Static emitter (`FuzzSchema → .baml source text`) → PR-C.
- Three-way oracle execution against mockllm → PR-B.
- Static prompt oracle → PR-C.
- Seed corpus → PR-C.
- CI integration job → PR-C.

## Touches codegen output

No. `go run ./cmd/regenerate-dynclient` is clean against `dynclient/`. Production codegen is byte-identical to master.

## Commits

1. `refactor(testharness)`: temp baml_src writer + replay artifacts + richer RunGoTest
2. `feat(issue340fuzz)`: shared FuzzSchema IR + OracleCase types + walker
3. `feat(issue340fuzz)`: rapid generators + value walker + invariant tests
4. `fix(bamlfuzz)`: rename + correct self-ref vs mutual-cycle contract + 12 review findings
5. `fix(bamlfuzz,testharness)`: close FindEnum + path-validation canonicalization bypasses
6. `fix(testharness)`: reject Windows drive-relative basenames in CheckReplayName + CheckBamlSrcName
7. `test(testharness)`: direct table-driven coverage for hasWindowsDrivePrefix

(Commits 2 and 3 reference the original `issue340fuzz` package name; commit 4 renames it to `bamlfuzz`. The CODE at HEAD uses `bamlfuzz` consistently — `git grep -nP 'issue340|F340' adapters/common/codegen/` is empty.)

## Verification

- `cd adapters/common && GOWORK=off go vet ./codegen/internal/{bamlfuzz,testharness}/...` ✅
- `cd adapters/common && GOWORK=off go test -race ./codegen/internal/{bamlfuzz,testharness}/... -count=1` ✅
- `cd adapters/common && GOWORK=off go test -race ./codegen/internal/bamlfuzz/... -count=3` ✅
- `cd adapters/common && GOWORK=off go test -race ./codegen/... -count=1` ✅ (PR #341 pool-lifecycle harness still passes)
- `go run ./cmd/regenerate-dynclient && git status dynclient/` ✅ clean
- `gofmt -l adapters/common/codegen/internal/{bamlfuzz,testharness}/` ✅ empty

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: adapters/common/codegen/internal/testharness/replay.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+// quadruple is serialized via this helper into a stable JSON file
+// the developer can rerun deterministically.
+func WriteReplayArtifact(t *testing.T, dir, name string, content json.RawMessage) {
+	t.Helper()
+	var buf bytes.Buffer
+	if err := json.Indent(&buf, content, "", "  "); err != nil {
+		t.Fatalf("indent replay artifact %s: %v\nraw: %s", name, err, string(content))
+	}
+	buf.WriteByte('\n')
+	WriteFile(t, filepath.Join(dir, name), buf.String())
```
[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/internal/testharness/replay.go:38]: @coderabbitai Extracted `CheckReplayName` as a pure function enforcing the documented basename contract: absolute paths, `..`, `.`, and any name containing a path separator are rejected with `t.Fatalf`. The function doc now explicitly says the name must be a basename. Pushed in `8162d0598`.

File: adapters/common/codegen/internal/testharness/runner.go [hunk 2]
```diff
@@ -28,16 +29,64 @@ func RunGoBuild(t *testing.T, dir string) {
...
+	if opts.Fuzz != "" {
+		args = append(args, "-fuzz", opts.Fuzz)
 	}
 	args = append(args, "./...")
 	cmd := exec.Command("go", args...)
 	cmd.Dir = dir
-	cmd.Env = append(os.Environ(), "GOWORK=off")
+	env := append(os.Environ(), "GOWORK=off")
+	env = append(env, opts.Env...)
+	cmd.Env = env
```
[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/internal/testharness/runner.go:103]: @coderabbitai When `opts.Fuzz != ""` the package target is now `"."` (single package, as required by `go help testflag`). `./...` is kept for the non-fuzz path. No iteration since fuzzing is defined for exactly one target. Pushed in `8162d0598`.

File: adapters/common/codegen/internal/bamlfuzz/generator.go [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+		flabel := label + "." + prop.Name
+		fv := c.drawValueForType(t, prop.Type, flabel)
+		fields[i] = FuzzFieldValue{Name: prop.Name, Value: fv}
+	}
+	return FuzzValue{
+		Kind:      KindClassRef,
+		ClassName: className,
+		Fields:    fields,
+	}
+}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/internal/bamlfuzz/generator.go:355]: @coderabbitai Added `cls, ok := c.schema.FindClass(className)` with `t.Fatalf` on `!ok` including `className` and the contextual label, so a generator IR/invariant violation surfaces immediately instead of producing an empty class-shaped value. Pushed in `8162d0598`.

File: adapters/common/codegen/internal/bamlfuzz/selfref.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+func collectClassRefs(t FuzzType, out map[string]bool) {
+	switch t.Kind {
+	case KindClassRef:
+		out[t.Ref] = true
+	case KindOptional, KindList, KindMap:
+		if t.Inner != nil {
+			collectClassRefs(*t.Inner, out)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/internal/bamlfuzz/selfref.go:115]: @coderabbitai Added a comment near the `KindMap` branch in `collectClassRefs` documenting the v1 invariant: map keys are always `KindString` (enforced at schema/generator construction), so class-ref reachability only needs to inspect `Inner`. Pushed in `8162d0598`.

File: adapters/common/codegen/internal/testharness/runner.go [hunk 5]
```diff
@@ -28,16 +29,64 @@ func RunGoBuild(t *testing.T, dir string) {
...
+		args = append(args, "-tags", opts.Tags)
+	}
+	if opts.Run != "" {
+		args = append(args, "-run", opts.Run)
+	}
+	if opts.Fuzz != "" {
+		args = append(args, "-fuzz", opts.Fuzz)
 	}
 	args = append(args, "./...")
 	cmd := exec.Command("go", args...)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/internal/testharness/runner.go:92]: @coderabbitai `RunGoTestArgs` now filters `opts.Env` to drop any entry whose key is `GOWORK` before appending, so `GOWORK=off` cannot be overridden by a caller. Final order: `os.Environ() + "GOWORK=off" + filtered(opts.Env)`. Pushed in `8162d0598`.

File: adapters/common/codegen/internal/testharness/baml_src.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+		t.Fatalf("mkdir baml_src: %v", err)
+	}
+	for name, content := range files {
+		path := filepath.Join(root, name)
+		if parent := filepath.Dir(path); parent != root {
+			if err := os.MkdirAll(parent, 0o755); err != nil {
+				t.Fatalf("mkdir %s: %v", parent, err)
+			}
+		}
+		WriteFile(t, path, content)
```
[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/internal/testharness/baml_src.go]: @coderabbitai Extracted `CheckBamlSrcName` as a pure testable function. It rejects absolute paths and any cleaned path that escapes the `baml_src/` root (`filepath.Rel` containment check). The helper now `t.Fatalf`s with the offending name before joining. Narrow defensive hardening, not a generic path sanitizer, since current callers pass deterministic generator-derived filenames. Pushed in `8162d0598`.

File: adapters/common/codegen/internal/issue340fuzz/normalize.go [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+		return WalkResult{}, err
+	}
+	if err := state.renderClassExpected(&expectBuf, value); err != nil {
+		return WalkResult{}, err
+	}
+	for cls, depth := range state.depths {
+		if depth > 0 {
+			meta.RecursionDepths[cls] = depth
+		}
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/internal/issue340fuzz/normalize.go]: @coderabbitai Peak recursion depth is now tracked at increment time (`s.meta.RecursionDepths[cls.Name] = max(...)` after the depth++); the post-walk copy loop was removed. New `TestWalkRecordsPeakRecursionDepth` locks `peak=2` for a 2-level self-ref so the previously-trivial assertion now actually verifies the bound. Pushed in `8162d0598`.

File: adapters/common/codegen/internal/testharness/replay.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/internal/testharness/replay.go:38]: @coderabbitai Acknowledged — your duplicate flag was correct: the narrowed `CheckReplayName` still cleans before checking, so `sub/../case.json` would normalize to `case.json` and slip through. Tightening it now to reject raw separators and `..`/`.` segments BEFORE `filepath.Clean`. Same canonicalization-before-validation pattern is being fixed in `CheckBamlSrcName` too (the `filepath.Rel` root-escape guard still held there, but the documented `..` rejection did not). Pushing in the next commit.

File: adapters/common/codegen/internal/testharness/replay.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+func CheckReplayName(name string) error {
+	if filepath.IsAbs(name) {
+		return fmt.Errorf("absolute name %q not allowed", name)
+	}
+	if name == "" || name == "." || name == ".." {
+		return fmt.Errorf("name must be a non-empty basename, got %q", name)
+	}
+	if strings.ContainsAny(name, `/\`) {
+		return fmt.Errorf("name must be a basename, got %q (contains path separator)", name)
+	}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/internal/testharness/replay.go:65]: @coderabbitai Good catch. Implementing the fix with an OS-independent `hasWindowsDrivePrefix` detector instead of `filepath.VolumeName` so the regression tests actually fire on our Linux/macOS CI (VolumeName is empty on non-Windows, so a test using it would silently skip). Applying the same guard to `CheckBamlSrcName` since it has the same top-level drive-relative gap. Pushing in the next commit.

[the remark this answers is not in the document]
Comment (invakid404) [reply adapters/common/codegen/internal/testharness/replay.go:65]: @coderabbitai Landed in `232ad3a8b` (the Windows drive-prefix guard) + `b7ff3681f` (direct table-driven tests for the `hasWindowsDrivePrefix` helper edge cases). Both `CheckReplayName` and `CheckBamlSrcName` now reject `C:case.json`, `D:foo`, `d:bar`, etc. The helper uses an ASCII-letter + colon check so the regression tests fire on every CI platform.
